--- gnulib/build-aux/do-release-commit-and-tag	2012-09-16 10:56:58.000000000 +0700
+++ gl/build-aux/do-release-commit-and-tag	2012-09-16 11:01:14.000000000 +0700
@@ -86,6 +86,10 @@
 branch=$(git branch | sed -ne '/^\* /{s///;p;q;}')
 builddir=.
 
+for gsort in $SORT gsort sort; do
+  echo 1|$gsort -V >/dev/null 2>/dev/null && break
+done
+
 while test $# != 0
 do
   # Handle --option=value by splitting apart and putting back on argv.
@@ -125,7 +129,7 @@
   || die 'failed to determine previous version number from .prev-version'
 
 # Verify that $ver is sensible (> .prev-version).
-case $(printf "$prev_ver\n$ver\n"|sort -V -u|tr '\n' ':') in
+case $(printf "$prev_ver\n$ver\n"|$gsort -V -u|tr '\n' ':') in
   "$prev_ver:$ver:") ;;
   *) die "invalid version: $ver (<= $prev_ver)";;
 esac
