diff --git a/build-aux/gitlog-to-changelog b/build-aux/gitlog-to-changelog
index e02d34c..aeae458 100755
--- a/build-aux/gitlog-to-changelog
+++ b/build-aux/gitlog-to-changelog
@@ -99,6 +99,8 @@ at the beginning of a log message line.  They are not copied to the output.
   Co-authored-by: Joe User <user\@example.com>
     List the specified name and email address on a second
     ChangeLog header, denoting a co-author.
+  Omit-from-ChangeLog: Yes
+    No ChangeLog entry is made for this commit.
   Signed-off-by: Joe User <user\@example.com>
     These lines are simply elided.
 
@@ -323,10 +325,14 @@ sub git_dir_option($)
         strftime ("%F", localtime ($1)), $2;
 
       my @coauthors = grep /^Co-authored-by:.*$/, @line;
+
+      my $omit = grep /^Omit-from-ChangeLog:\s+[Yy]es$/, @line;
+
       # Omit meta-data lines we've already interpreted.
       @line = grep !/^(?:Signed-off-by:[ ].*>$
                        |Co-authored-by:[ ]
                        |Copyright-paperwork-exempt:[ ]
+                       |Omit-from-ChangeLog:[ ]
                        )/x, @line;
 
       # Remove leading and trailing blank lines.
@@ -351,33 +357,33 @@ sub git_dir_option($)
               . substr ($_, 5) . "\n";
         }
 
-      # If clustering of commit messages has been disabled, if this header
-      # would be different from the previous date/name/email/coauthors header,
-      # or if this or the previous entry consists of two or more paragraphs,
-      # then print the header.
-      if ( ! $cluster
-          || $date_line ne $prev_date_line
-          || "@coauthors" ne "@prev_coauthors"
-          || $multi_paragraph
-          || $prev_multi_paragraph)
-        {
-          $prev_date_line eq ''
-            or print "\n";
-          print $date_line;
-          @coauthors
-            and print join ("\n", @coauthors), "\n";
-        }
-      $prev_date_line = $date_line;
-      @prev_coauthors = @coauthors;
-      $prev_multi_paragraph = $multi_paragraph;
-
       # If there were any lines
       if (@line == 0)
         {
           warn "$ME: warning: empty commit message:\n  $date_line\n";
         }
-      else
+      elsif ( ! $omit )
         {
+          # If clustering of commit messages has been disabled, if this header
+          # would be different from the previous date/name/email/coauthors header,
+          # or if this or the previous entry consists of two or more paragraphs,
+          # then print the header.
+          if ( ! $cluster
+              || $date_line ne $prev_date_line
+              || "@coauthors" ne "@prev_coauthors"
+              || $multi_paragraph
+              || $prev_multi_paragraph)
+            {
+              $prev_date_line eq ''
+                or print "\n";
+              print $date_line;
+              @coauthors
+                and print join ("\n", @coauthors), "\n";
+            }
+          $prev_date_line = $date_line;
+          @prev_coauthors = @coauthors;
+          $prev_multi_paragraph = $multi_paragraph;
+
           if ($append_dot)
             {
               # If the first line of the message has enough room, then
